build/e2e_upgrade_test.sh

```diff
@@ -0,0 +1,196 @@
+#!/usr/bin/env bash
+
+# Copyright 2025 Google LLC All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+set -o pipefail
+
+BASE_VERSION=$1
+PROJECT_ID=$2
+logBucketName="upgrade-test-container-logs"
+
+apt-get update && apt-get install -y jq
+export SHELL="/bin/bash"
+export KUBECONFIG="/root/.kube/config"
+mkdir -p /go/src/agones.dev/ /root/.kube/
+ln -s /workspace /go/src/agones.dev/agones
+cd /go/src/agones.dev/agones/test/upgrade
+
+# --- Function to print failure logs ---
+print_failure_logs() {
+    local cluster_name=$1
+    echo "ERROR: Upgrade test failed on cluster: $cluster_name"
+
+    job_pods=$(kubectl get pods -l job-name=upgrade-test-runner -o jsonpath="{.items[*].metadata.name}")
+    kubectl logs --tail=20 "$job_pods" || echo "Unable to retrieve logs for pod: $job_pods"
+    for pod in $job_pods; do
+        containers=$(kubectl get pod "$pod" -o jsonpath='{.spec.containers[*].name}')
+        for container in $containers; do
+            if [[ "$container" == "sdk-client-test" || "$container" == "upgrade-test-controller" ]]; then
+                echo "----- Logs from pod: $pod, container: $container -----"
+                kubectl logs "$pod" -c "$container" || echo "Failed to retrieve logs from $pod/$container"
+            fi
+        done
+    done
+
+    log_url="https://console.cloud.google.com/logs/query;storageScope=storage,projects%2F${PROJECT_ID}%2Flocations%2Fglobal%2Fbuckets%2Fupgrade-test-container-logs%2Fviews%2F_AllLogs;query=$(printf 'resource.labels.container_name=(\"upgrade-test-controller\" OR \"sdk-client-test\")' | jq -sRr @uri);cursorTimestamp=$(date -u +%Y-%m-%dT%H:%M:%SZ);duration=PT10M?project=${PROJECT_ID}"
+    echo "Logs from log bucket: $log_url"
+}
```

Comment:
LGTM, but I just got a question, are we sure this `print_failure_logs` will not only show logs from the last kube context from this ? 
```
for cloudProduct in "${cloudProducts[@]}"
do
    for version in "${!versionsAndRegions[@]}"
```

I was thinking to re-do the `    gcloud container clusters get-credentials "$testCluster" --region="$testClusterLocation" --project="$PROJECT_ID"` or using the right context before calling this function, it doesn't seems to be done from L162-L194

Reply:
My comment is not really clear, but I'm referring to the calls of `kubectl logs --tail ..` which might use the context from the last call of `gcloud container clusters get-credentials ..`, as we were previously only storing the PID and the log file, then waiting for pid to be done to just print / logs it without kubectl calls

But with this, it might calls kubectl using the context from the last one of the list (which might not have errors / not the expected context)

If that make sense 😄

Reply:
That part looks all good with the wait, but about this one:

for pid in "${pids[@]}"; do
    ...
    print_failure_logs ..
    ...
done

Feels that the calls to print_failure_logs from this loop will use the same kube context though, compared to how it was logged before 

L183 and L191